source/RenderBuffer.cpp

```diff
@@ -0,0 +1,299 @@
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "RenderBuffer.h"
+
+#include "Logger.h"
+#include "Screen.h"
+#include "Shader.h"
+
+#include "opengl.h"
+
+namespace {
+	Shader shader;
+	GLint sizeI = -1;
+	GLint positionI = -1;
+	GLint scaleI = -1;
+	GLint srcpositionI = -1;
+	GLint srcscaleI = -1;
+
+	GLint texI = -1;
+
+	GLuint vao = -1;
+	GLuint vbo = -1;
+}
+
+// Initialize the shaders.
+void RenderBuffer::Init()
+{
+	static const char *vertexCode =
+		"// vertex blit shader\n"
+		"precision mediump float;\n"
+		"uniform vec2 size;\n"
+		"uniform vec2 position;\n"
+		"uniform vec2 scale;\n"
+
+		"uniform vec2 srcposition;\n"
+		"uniform vec2 srcscale;\n"
+
+		"in vec2 vert;\n"
+		"out vec2 tpos;\n"
+
+		"void main() \n"
+		"{\n"
+		"  gl_Position = vec4((position + vert * size) * scale, 0, 1);\n"
+		"  vec2 tvert = vert + vec2(.5, .5);\n" // Convert from vertex to texture coordinates.
+		"  vec2 tsize = size * srcscale;\n"    // Convert from screen to texture coordinates.
+		"  vec2 tsrc = srcposition * srcscale;\n" // Convert from screen to texture coordinates.
+		"  tpos = tvert * tsize + tsrc;\n"
+		"  tpos.y = 1.0 - tpos.y;\n" // Negative is up.
+		"}\n";
+
+	static const char *fragmentCode =
+		"// fragment blit shader\n"
+		"precision mediump float;\n"
+		"precision mediump sampler2D;\n"
+		"uniform sampler2D tex;\n"
+
+		"in vec2 tpos;\n"
+		"out vec4 finalColor;\n"
+
+		"void main() {\n"
+		"  if(tpos.x > 0.0 && tpos.y > 0.0 &&\n"
+		"      tpos.x < 1.0 && tpos.y < 1.0 )\n"
+		"    finalColor = texture(tex, tpos);\n"
+		"  else\n"
+		"    discard;\n"
+		"}\n";
+
+	shader = Shader(vertexCode, fragmentCode);
+	sizeI = shader.Uniform("size");
+	positionI = shader.Uniform("position");
+	scaleI = shader.Uniform("scale");
+	srcpositionI = shader.Uniform("srcposition");
+	srcscaleI = shader.Uniform("srcscale");
+	texI = shader.Uniform("tex");
+
+	// Generate the vertex data for drawing sprites.
+	glGenVertexArrays(1, &vao);
+	glBindVertexArray(vao);
+
+	glGenBuffers(1, &vbo);
+	glBindBuffer(GL_ARRAY_BUFFER, vbo);
+
+	GLfloat vertexData[] = {
+		-.5f, -.5f,
+		-.5f,  .5f,
+		.5f, -.5f,
+		.5f,  .5f
+	};
+	glBufferData(GL_ARRAY_BUFFER, sizeof(vertexData), vertexData, GL_STATIC_DRAW);
+
+	glEnableVertexAttribArray(shader.Attrib("vert"));
+	glVertexAttribPointer(shader.Attrib("vert"), 2, GL_FLOAT, GL_FALSE, 2 * sizeof(GLfloat), nullptr);
+
+	// unbind the VBO and VAO
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
+}
+
+
+
+// Create a texture of the given size that can be used as a render target
+RenderBuffer::RenderBuffer(const Point& dimensions):
+	size(dimensions)
+{
+	// Generate a framebuffer, and bind it.
+	glGenFramebuffers(1, &framebuffer);
+	glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
+
+	// Generate the texture.
+	glGenTextures(1, &texid);
+	glBindTexture(GL_TEXTURE_2D, texid);
```

Comment:
It would be nice if this was a texture 2D array so that it could be plugged into the sprite shader

Reply:
I'm inclined to leave this for a future update if needed. I've tried a couple of times to get this working, and I keep getting errors from the framebuffer.